ledger/evaldebugger.go

```diff
@@ -0,0 +1,66 @@
+// Copyright (C) 2019-2022 Algorand, Inc.
+// This file is part of go-algorand
+//
+// go-algorand is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// go-algorand is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package ledger
+
+import (
+	"fmt"
+
+	"github.com/algorand/go-algorand/config"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/bookkeeping"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/ledger/internal"
+	"github.com/algorand/go-algorand/ledger/ledgercore"
+)
+
+// DebuggerLedgerForEval is a ledger interface for the debugger.
+type DebuggerLedgerForEval interface {
+	internal.LedgerForEvaluator
+	Latest() basics.Round
+}
+
+// EvalForDebugger processes a transaction group for the debugger.
+func EvalForDebugger(l DebuggerLedgerForEval, stxns []transactions.SignedTxn) (ledgercore.StateDelta, []transactions.SignedTxnInBlock, error) {
```

Comment:
I think this should be `EvalWithDebugging`.  We often call interfaces XxxForYyy when it is a narrowing of Xxx, such that it is usable for the purposes of Yyy. Like "LedgerForLogic".

But my larger question is why we need to this function and `ProcessTransactionGroupForDebugger` at all.  We should be trying to construct an object that "fools" the normal block evalautor into working on our `SimulationLedger`. That is, make a `SimulationLedger` that conforms to the interface that StartEvaluator needs (a LedgerForEvaluator`) and just hand it in and let it do its job.  While I recognize that LedgerForEvaluator is an internal interface, that doesn't matter in Go.  If we make an object that conforms to it, a `SimulationLedger`, the code there will be happy.

Our goal should be avoid writing this kind of wrapper code (the two *ForDebugger routines). We will inevitably end up processing transactions wrong. Instead, teach the existing evaluator to do what we need (call some debugging hooks?). I speculate that the way to tell the evaluator to call such hooks is by making some otherwise unused flags in the ConsensusParam structure.  So the evaluator says, "Oh, I'm running with proto.Debug? Then I'll create my internal evalParams with Trace set".

Reply:
If you push your handler code into NodeInterface, you'll have access to real, honest to god, `*data.Ledger` and you won't have to wrestle with LedgerForAPI.